fdb/fdb.go

```diff
@@ -0,0 +1,236 @@
+package fdb
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/apple/foundationdb/bindings/go/src/fdb"
+	"github.com/apple/foundationdb/bindings/go/src/fdb/directory"
+	"github.com/apple/foundationdb/bindings/go/src/fdb/tuple"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipni/dhstore"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+	"lukechampine.com/blake3"
+)
+
+var (
+	_ dhstore.DHStore = (*FDBDHStore)(nil)
+
+	logger        = logging.Logger("store/fdb")
+	fdbHasherPool sync.Pool
+)
+
+const (
+	fdbMaxValueBytes      = 100_000 // 100 KB
+	fdbBlake3HashLength   = 32
+	fdbMultihashDirectory = "mh"
+	fdbMetadataDirectory  = "md"
+)
+
+type FDBDHStore struct {
+	db fdb.Database
+}
+
+func init() {
+	fdbHasherPool.New = func() any {
+		return blake3.New(fdbBlake3HashLength, nil)
+	}
+}
+
+func NewFDBDHStore(o ...Option) (*FDBDHStore, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+	if err := fdb.APIVersion(opts.apiVersion); err != nil {
+		return nil, err
+	}
+	db, err := fdb.OpenDatabase(opts.clusterFile)
+	if err != nil {
+		panic(err)
+	}
+	return &FDBDHStore{
+		db: db,
+	}, nil
+}
+
+func (f *FDBDHStore) MergeIndex(mh multihash.Multihash, vk dhstore.EncryptedValueKey) error {
+	dmh, err := multihash.Decode(mh)
+	if err != nil {
+		return dhstore.ErrMultihashDecode{Err: err, Mh: mh}
+	}
+	if multicodec.Code(dmh.Code) != multicodec.DblSha2_256 {
+		return dhstore.ErrUnsupportedMulticodecCode{Code: multicodec.Code(dmh.Code)}
+	}
+	if len(vk) > fdbMaxValueBytes {
+		return fmt.Errorf("value key cannot be larger than 100 KB, got: %d", len(vk))
+	}
+	_, err = f.db.Transact(func(transaction fdb.Transaction) (any, error) {
+		// Hash encrypted value key to fix the size of key tuple. It may be
+		// more efficient to store the encrypted value key as is if it's
+		// short enough. But hashing it would offer a more even utilisation
+		// of the cluster by eliminating potential common prefixes in the
+		// encrypted value key.
+		// TODO: investigate if there are benefits here to be explored in
+		//       terms of reduction in disk utilisation.
+		vkh, err := f.hash(vk)
```

Comment:
The original key should be short enough without second layer of hashing. We also know that it's aways 32 bytes (sha256)

Reply:
Sure. Do we have a max expected size for `metadata` key? I believe those are not necessarily multihashes. I would like to enforce max supported in fdb since it impacts performance. Otherwise, i'll keep the hashing for metadata keys. 

Thoughts?

Reply:
Got it, but it's not a multihash right? 

Going with 32 bytes max length enforcement.

Reply:
It's not a multihash, just fixed 32 byte array